drivers/modem/quectel-bg9x.c

```diff
@@ -0,0 +1,1155 @@
+	char   sendbuf[sizeof("AT+QIRD=##,####")] = {0};
+	int    ret;
+	struct socket_read_data sock_data;
+
+	/* Modem command to read the data. */
+	struct modem_cmd data_cmd[] = { MODEM_CMD("+QIRD: ", on_cmd_sock_readdata, 0U, "") };
+
+	if (!buf || len == 0) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	if (flags & ZSOCK_MSG_PEEK) {
+		errno = ENOTSUP;
+		return -1;
+	}
+
+	snprintk(sendbuf, sizeof(sendbuf), "AT+QIRD=%d,%d", sock->sock_fd, len);
+
+	/* Socket read settings */
+	(void) memset(&sock_data, 0, sizeof(sock_data));
+	sock_data.recv_buf     = buf;
+	sock_data.recv_buf_len = len;
+	sock_data.recv_addr    = from;
+	sock->data	       = &sock_data;
+	mdata.sock_fd	       = sock->sock_fd;
+
+	/* Tell the modem to give us data (AT+QIRD=sock_fd,data_len). */
+	ret = modem_cmd_send(&mctx.iface, &mctx.cmd_handler,
+			     data_cmd, ARRAY_SIZE(data_cmd), sendbuf, &mdata.sem_response,
+			     MDM_CMD_TIMEOUT);
+	if (ret < 0) {
+		errno = -ret;
+		ret = -1;
+		goto exit;
+	}
+
+	/* HACK: use dst address as from */
+	if (from && fromlen) {
+		*fromlen = sizeof(sock->dst);
+		memcpy(from, &sock->dst, *fromlen);
+	}
+
+	/* return length of received data */
+	errno = 0;
+	ret = sock_data.recv_read_len;
+
+exit:
+	/* clear socket data */
+	sock->data = NULL;
+	return ret;
+}
+
+/* Func: offload_read
+ * Desc: This function reads data from the given socket object.
+ */
+static ssize_t offload_read(void *obj, void *buffer, size_t count)
+{
+	return offload_recvfrom(obj, buffer, count, 0, NULL, 0);
+}
+
+/* Func: offload_write
+ * Desc: This function writes data to the given socket object.
+ */
+static ssize_t offload_write(void *obj, const void *buffer, size_t count)
+{
+	return offload_sendto(obj, buffer, count, 0, NULL, 0);
+}
+
+/* Func: offload_poll
+ * Desc: This function polls on a given socket object.
+ */
+static int offload_poll(struct zsock_pollfd *fds, int nfds, int msecs)
+{
+	int i;
+	void *obj;
+
+	/* Only accept modem sockets. */
+	for (i = 0; i < nfds; i++) {
+		if (fds[i].fd < 0) {
+			continue;
+		}
+
+		/* If vtable matches, then it's modem socket. */
+		obj = z_get_fd_obj(fds[i].fd,
+				   (const struct fd_op_vtable *) &offload_socket_fd_op_vtable,
+				   EINVAL);
+		if (obj == NULL) {
+			return -1;
+		}
+	}
+
+	return modem_socket_poll(&mdata.socket_config, fds, nfds, msecs);
+}
+
+/* Func: offload_ioctl
+ * Desc: Function call to handle various misc requests.
+ */
+static int offload_ioctl(void *obj, unsigned int request, va_list args)
+{
+	switch (request) {
+	case ZFD_IOCTL_POLL_PREPARE:
+		return -EXDEV;
+
+	case ZFD_IOCTL_POLL_UPDATE:
+		return -EOPNOTSUPP;
+
+	case ZFD_IOCTL_POLL_OFFLOAD:
+	{
+		/* Poll on the given socket. */
+		struct zsock_pollfd *fds;
+		int nfds, timeout;
+
+		fds = va_arg(args, struct zsock_pollfd *);
+		nfds = va_arg(args, int);
+		timeout = va_arg(args, int);
+
+		return offload_poll(fds, nfds, timeout);
+	}
+
+	default:
+		errno = EINVAL;
+		return -1;
```

Comment:
`return -EINVAL` will work?

Reply:
It will work, but I think the standard posix behavior is to return errors in `errno` variable, though this function doesn't seem to be POSIX compliant.